src/lib/Microsoft.Health.Fhir.R4.Ingest/Service/R4FhirImportService.cs

```diff
@@ -72,26 +72,66 @@ public virtual async Task<string> SaveObservationAsync(ILookupTemplate<IFhirTemp
                 existingObservation = await GetObservationFromServerAsync(identifier).ConfigureAwait(false);
             }
 
-            Model.Observation result;
+            Model.Observation result = null;
+            ResourceOperation operationType = ResourceOperation.NoOperation;
+
             if (existingObservation == null)
             {
                 var newObservation = GenerateObservation(config, observationGroup, identifier, ids);
                 result = await _client.CreateAsync(newObservation).ConfigureAwait(false);
-                _logger.LogMetric(IomtMetrics.FhirResourceSaved(ResourceType.Observation, ResourceOperation.Created), 1);
+                operationType = ResourceOperation.Created;
             }
             else
             {
+                Model.Observation mergedObservation = null;
+
                 var policyResult = await Policy<Model.Observation>
-                    .Handle<FhirOperationException>(ex => ex.Status == System.Net.HttpStatusCode.Conflict || ex.Status == System.Net.HttpStatusCode.PreconditionFailed)
-                    .RetryAsync(2, async (polyRes, attempt) =>
-                    {
-                        existingObservation = await GetObservationFromServerAsync(identifier).ConfigureAwait(false);
-                    })
-                    .ExecuteAndCaptureAsync(async () =>
-                    {
-                        var mergedObservation = MergeObservation(config, existingObservation, observationGroup);
-                        return await _client.UpdateAsync(mergedObservation, versionAware: true).ConfigureAwait(false);
-                    }).ConfigureAwait(false);
+                     .Handle<FhirOperationException>(ex => ex.Status == System.Net.HttpStatusCode.Conflict || ex.Status == System.Net.HttpStatusCode.PreconditionFailed)
+                     .RetryAsync(2, async (polyRes, attempt) =>

```

Comment:
Any particular reason for splitting the logic here between the `RetryAsync` and `ExecuteAndCaptureAsync`? I believe that `ExecuteAndCaptureAsync` runs inside of the policy created by `RetryAsync`, so I'm curious what the benefit is for the split. Personally, it makes the logic a bit hard to follow.

Reply:
RetryAsync only executes on failure.  ExecuteAndCapture runs in every every attempt.

So flow is as such.  First attempt is Execute block.  If there is no failure we are done.   If there is a failure the retry block executes to refresh the observation and then we execute the ExecuteAndCapture step again.  Personally I think this makes the logic easier to follow.  We have code we always execute in an attempt and then we have specific code on an old e-tag that refreshes our observation.  The other benefit is in the future if there are other errors codes that require different retry steps they are easier integrated into the current logic.

Reply:
This code has been here for a while, but I believe this is a pattern outlined by [Polly](https://github.com/App-vNext/Polly#post-execution-capturing-the-result-or-any-final-exception).

The ExecuteAndCapture block will always run first. If there is an exception the RetryAsync block runs (to try and recover from the issue)  before the ExecuteAndCapture block runs again. I personally like the syntax, maybe I can make the code easier to read.

Reply:
I like it too Nate but I am biased an wrote it originally.  The key thing about this pattern too is it extends nicely if you have different retry blocks for different exception conditions in the future.

Reply:
Thanks for the clarification. The split of logic between the two methods + the order of execution wasn't clear but now I follow.